dlf/dlflogstore.man
.lf 1 dlflogstore.man
.\" @(#)$RCSfile: dlflogstore.man,v $ $Revision: 1.1 $ $Date: 2003/08/20 13:06:12 $ CERN IT-ADC Vitaly Motyakov
.\" Copyright (C) 2003 by CERN/IT/ADC
.\" All rights reserved
.\"
.TH DLFLOGSTORE 1 "$Date: 2003/08/20 13:06:12 $" CASTOR "DLF Administrator Commands"
.SH NAME
dlflogstore \- store log messages from the given logfile into the CASTOR
DLF database
.SH SYNOPSIS
.B dlflogstore
.BI -f " file_name"
.BI -h " host_name"
[
.BI -p " port_number"
] [
.BI -s " severity_level"
]
.SH DESCRIPTION
.B dlflogstore
stores log messages from the given logfile into the CASTOR
DLF database.
.TP
.I file_name
is the name of the input log file. It must be at most
CA_MAXPATHLEN characters long.
.TP
.I host_name
is the name of the DLF server host.
.TP
.I port_number
is the TCP port number on which the DLF server is listening. If not specified
the default port will be used.
.TP
.I severity_level
is the name of the severity level. If specified only messages with this
severity level will be send to the DLF server for storing.
.SH EXAMPLE
.nf
.ft CW
dlflogstore -f /usr/spool/stage/log -h castor.cern.ch
.ft
.fi
.SH EXIT STATUS
This program returns 0 if the operation was successful or >0 if the operation
failed.
.SH AUTHOR
\fBCASTOR\fP Team <castor.support@cern.ch>


CHANGELOG.md
## 5.11.23.0-1 (2026-09-25)

### New Features

- [frontend] Add user configuration improvements for cta-admin-grpc (cta/CTA!1635)

## 5.11.22.0-1 (2026-09-04)

### New Features

- [frontend] Implement JWT blacklisting on the CTA Frontend (cta/CTA!1548)

- [runtime] Add additional validation logic to application config (cta/CTA!1567)


### Bug Fixes

- [scheduler] Fix issues after pg1 release rest: repack admin commands, archive...

- [Tools] Fix 'priority' column in 'cta-admin dr ls' command (cta/CTA!1546)

- [runtime] Make signals, telemetry and health handling more robust (cta/CTA!1554)

- [runtime] Add bounds checking and better error diagnostics to TOML loading (cta/CTA!1559)

## 5.11.21.0-1 (2026-08-25)

### New Features

- [scheduler] Add scheduler workflow metrics to OpenTelementry (cta/CTA!1387)


### Changes

- [Tools] Move cta-admin af ch command to tf ch (cta/CTA!1525)
- [Tools] Replace cta-objectstore-tools.conf with cta-scheduler.conf (cta/CTA!1540)
- [maintd] Change scheduler.objectstore_backend_path to scheduler.config_file (cta/CTA!1518)


### Bug Fixes

- [frontend] Fix logic bug in frontend authentication and SEGFAULT in cta-admin (cta/CTA!1558)
- [Tools] Fix cta-verify-file (cta/CTA!1545)
- [Tools] Add `--storage-class` option to `cta-send-event` (cta/CTA!1542)


### Performance Improvements

- [taped] Skip block reposition request if drive is already positioned correctly (cta/CTA!1527)

## 5.11.20.0-1

### New Features

- [Tools] Add optional storage class statistics, per tape, to cta-admin tool (cta/CTA!1410)
- [Misc] Add event_name to important CTA log messages (cta/CTA!1381)

### Changes

- [Tools] Change short option for cta-admin af ch --idfile from -I to -F (cta/CTA!1453)
- [taped] Taped no longer sets its own user, group and SYS_RAWIO capability (cta/CTA!1426)
- [Misc] Improve .service and .logrotate configuration (cta/CTA!1524)
- [Misc] Modernise RPM spec metadata and file handling (cta/CTA!1507)
- [Misc] Move systemd units to `/usr/lib/systemd/system` (cta/CTA!1497)
- [Misc] Change systemd to weak RPM dependency of CTA (cta/CTA!1509)
- [Misc] Change logrotate to weak RPM dependency of CTA (cta/CTA!1505)
- [Misc] Remove RPM changelog; package CHANGELOG.md as doc (cta/CTA!1508)
- [Misc] Remove noreplace from example configuration files (cta/CTA!1506)
- [taped] Drive process crashing produces CRIT message instead of INFO (cta/CTA!1499)
- [taped] cta-eosdf no longer sources sysconfig (cta/CTA!1455)

### Bug Fixes

- [frontend] Forbid reclaiming tapes with recently deleted files (cta/CTA!1418)
- [frontend] PREPARE gRPC request shouldn't fail without a storage class (cta/CTA!1471)
- [taped] Fix crash due to accidental pointer arithmetic in exception msg (cta/CTA!1502)
- [taped] Fix opentelemetry shutdown race condition with objectstore (cta/CTA!1500)
- [scheduler]: Fix archive unmount decision logic for postgres scheduler (cta/CTA!1450)
- [frontend] Fix frontend catalogue cache for mount rules (cta/CTA!1490)
- [CI] Ensure CTA images trust CERN Root CA (cta/CTA!1535)
- [Misc] Ensure CMake correctly installs executables as PROGRAMS (cta/CTA!1523)

### Security

- [CI] Generate SBOM for each Docker image (cta/CTA!1532)

### Other

- [CI] Add system tests to cover staging tokens on EOS-CTA (cta/CTA!1493)

## 5.11.19.0-1

### Changes

- [frontend] Split Admin/WFE frontends, add mTLS (cta/CTA!1367)
- [catalogue] Forbid adding or changing encryption key of non-empty tapes (cta/CTA!1414)

### Deprecations

- [frontend] Deprecate storage class field in EOS-CTA xattrs (cta/CTA!1424)

### Other

- [CI] Add Enstore and EnstoreLarge tape tests (cta/CTA!1412)

## 5.11.18.0-2

### Bug Fixes

- [Misc] Update required runtime version of opentelemetry-cpp to v1.26.0 (cta/CTA!1370)

## 5.11.18.0-1

### New Features

- [maintd] Add catalogue and telemetry conf files to runtime directory (cta/CTA!1303)
- [Misc] Package basic JSON logging schema (cta/CTA!1304)
- [Misc] Add source location and CTA version to log lines (cta/CTA!1348)

### Changes

- [frontend] Allow admin commands or physics workflows to be disabled on CTA frontend (cta/CTA!1336)
- [Misc] Replace reflect-cpp library with custom implementation (cta/CTA!1291)
- [Misc] Change Catalogue backend install behaviour (cta/CTA!1295)

### Bug Fixes

- [catalogue] Improve CTA resilience to catalogue DB outages during data fetching (cta/CTA!1334)
- [taped] Improve readExactBlock error message contents (cta/CTA!1311)
- [taped] Enabling reading of tapes with misformed OSM tape format (cta/CTA!1343)
- [frontend] Prevent 'cta-admin dr rm' command from crashing the cta-frontend (cta/CTA!1305)
- [frontend] Remove race conditions in NegotiationRequestHandler (cta/CTA!1310)
- [frontend] Fix broken cta-admin fr ls command (cta/CTA!1340)
- [Misc] Fix compile issues when building CTA in Release mode (cta/CTA!1207)

### Performance Improvements

- [CI] Use python xrootd persistent clients for archive instead of xrdcp --parallel (cta/CTA!1274)

### Other

- [Misc] Update opentelemetry-cpp to v1.26.0 (cta/CTA!1321)


## 5.11.17.0-1

### New Features

- [maintd] Add CLI option to use runtime directory for app metadata (cta/CTA!1266)

### Changes

- [maintd] Update maintd to use TOML configuration (cta/CTA!1246)
- [maintd] Use declarative config to initialise OpenTelemetry SDK (cta/CTA!1248)
- [maintd] Add use_unix_domain_socket configuration entry for health server (cta/CTA!1282)
- [maintd] Remove daemon_user and daemon_group settings (cta/CTA!1252)

### Bug Fixes

- [frontend] Fix crash in 'cta-admin dr ls' command when using multiple...
- [frontend] Add error message when trying to change media type of non-empty tape (cta/CTA!1253)

### Removals

- [frontend] Stop using TAPE_POOL.SUPPLY column (cta/CTA!1245)

### Performance Improvements

- [taped] Optimise DriveState queries (cta/CTA!1198)

### Other

- [Misc] Rename *.conf.example files to *.example.conf (cta/CTA!1279)

## 5.11.16.1-1

### Bug Fixes


dlf/dlflogstore.c
/*
 * 
 * Copyright (C) 2003 by CERN/IT/ADC
 * All rights reserved
 *
 */

#ifndef lint
static char sccsid[] = "@(#)$RCSfile: dlflogstore.c,v $ $Revision: 1.3 $ $Date: 2003/11/03 08:54:23 $ CERN IT-ADC/CA Vitaly Motyakov";
#endif /* not lint */

#include <errno.h>
#include <grp.h>
#include <pwd.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>
#if defined(_WIN32)
#include <winsock2.h>
#else
#include <netinet/in.h>
#include <netdb.h>
#endif
#include "net.h"
#include "marshall.h"
#include "serrno.h"
#include "dlf.h"
#include "dlf_api.h"
#include "Cnetdb.h"
#include <ctype.h>
#if !defined(_WIN32)
#include <unistd.h>
#endif

#define	  DLF_KEY_DATE   1
#define	  DLF_KEY_HOST   2
#define   DLF_KEY_LVL    3
#define   DLF_KEY_PRGN   4
#define   DLF_KEY_PROG   5
#define   DLF_KEY_PID    6
#define   DLF_KEY_TID    7
#define	  DLF_KEY_MSGN   8
#define	  DLF_KEY_MSG    9
#define	  DLF_KEY_RQID  10
#define	  DLF_KEY_NSINV 11
#define	  DLF_KEY_SRQID 12
#define	  DLF_KEY_TPVID 13

int DLL_DECL dlf_get_pair _PROTO((FILE*, char*, char*));
int DLL_DECL dlf_iskeyword _PROTO((const char*));
int DLL_DECL dlf_send_log_message _PROTO((dlf_log_dst_t*, dlf_log_message_t*, int));

char *getconfent();
char *getenv();

struct _dlf_keyword {
	  int key_id;
	  char *key_name;
};

static struct _dlf_keyword keywords[] = {
	  { DLF_KEY_DATE, "DATE" },
	  { DLF_KEY_HOST, "HOST" },
	  { DLF_KEY_LVL,  "LVL" },
	  { DLF_KEY_PRGN, "PRGN" },
	  { DLF_KEY_PROG, "PROG" },
	  { DLF_KEY_PID,  "PID" },
	  { DLF_KEY_TID,  "TID" },
	  { DLF_KEY_MSGN, "MSGN" },
	  { DLF_KEY_MSG,  "MSG" },
	  { DLF_KEY_RQID, "RQID" },
	  { DLF_KEY_NSINV,"NSINV" },
	  { DLF_KEY_SRQID,"DLF.SRQID*" },
	  { DLF_KEY_TPVID,"DLF.TPVID*" },
	  {             0,"" } /* End of list */
};

extern struct _dlf_level g_dlf_levels[];

extern	char	*optarg;
extern	int	optind;

int main(argc, argv)
     int argc;
     char **argv;
{
	int errflg;
	char c;
	char file_name[CA_MAXPATHLEN + 1];
	uid_t uid;
	gid_t gid;
	dlf_log_message_t log_message;
	FILE *in_file;
	int rv;
	int gp_rv;
	char par_name[DLF_MAXPARNAMELEN + 1];
	char par_str_val[DLF_MAXSTRVALLEN + 1];
	char *p;
	int n;
	int key;
	int i;
	Cuuid_t sub_request_id;
	char *end_ptr;
	HYPER i64val;
	double dval;
	dlf_log_dst_t dst_host;
	int file_given = 0;
	int host_given = 0;
	int store_severity = -1;
	int port_number = 0;
	int new_line = 0;
	int last_message;
	struct servent *sp;
	int line_no = 1;

	uid = geteuid();
	gid = getegid();

#if defined(_WIN32)
	if (uid < 0 || gid < 0) {
		fprintf (stderr, DLF53);
		exit (USERR);
	}
#endif
	errflg = 0;
        while ((c = getopt (argc, argv, "f:s:h:p:?")) != EOF) {
                switch (c) {
		case 'f':
			if (strlen (optarg) <= CA_MAXPATHLEN) {
				strcpy (file_name, optarg);
				file_given++;
			}
			else {
				fprintf (stderr, "Illegal file name: %s\n", strerror(EINVAL));
				errflg++;
			}
                        break;
		case 's':
		        for (i = 0; g_dlf_levels[i].lvl_id != 0 && strcasecmp(optarg, g_dlf_levels[i].lvl_name) != 0 ; i++);
			store_severity = g_dlf_levels[i].lvl_id;
			if (store_severity == 0) {
			        fprintf (stderr, "Unrecognized severity level: %s\n", optarg);
				errflg++;
			}
			break;
		case 'h':
			if (strlen (optarg) <= CA_MAXHOSTNAMELEN) {
				strcpy (dst_host.name, optarg);
				host_given++;
			}
			else {
				fprintf (stderr, "Illegal host name: %s\n", strerror(EINVAL));
				errflg++;
			}
                        break;
		case 'p':
		        if (dlf_isinteger(optarg)) {
 				port_number = atoi (optarg);
			}
			else {
				fprintf (stderr, "Illegal port: %s\n", strerror(EINVAL));
				errflg++;
			}
                        break;
		case '?':
                        errflg++;
                        break;
                default:
		        errflg++;
                        break;
		}
	}
        if (optind < argc) {
                errflg++;
	}
	if (!(file_given && host_given)) {
	        fprintf (stderr, "Not all required parameters are given\n");
	        errflg++;
	}
        if (errflg) {
                fprintf (stderr, "usage: %s %s", argv[0],
		    "-f file_name -h host_name [-p port_number] [-s severity_level]\n");
                exit (USERR);
	}
#if defined(_WIN32)
	if (WSAStartup (MAKEWORD (2, 0), &wsadata)) {
		fprintf (stderr, DLF52);
		exit (SYERR);
	}
#endif
	/* Set port */
	if (port_number == 0) {
	  if ((p = getenv ("DLF_PORT")) || (p = getconfent ("DLF", "PORT", 0))) {
	        port_number = atoi (p);
	  }
	  else if ((sp = Cgetservbyname ("dlf", "tcp")) != NULL) {
		port_number = ntohs(sp->s_port);
	  }
	  else {
	        port_number = DLF_PORT;
	  }
	}


	/* Initialize dst_host structure */
	dst_host.dst_type = DLF_DST_TCPHOST;
	dst_host.descr = -1; /* Not connected yet */
	dst_host.port = port_number;

	log_message.param_list.head = NULL;
	log_message.param_list.tail = NULL;
	log_message.next_tapevid_idx = 1;
	log_message.next_subreq_idx = 1;

	/* Open the file for reading */
	in_file = fopen(file_name, "r");
	if (in_file == NULL) {
	  fprintf (stderr, "Error opening log file %s: %s\n", file_name, sstrerror(errno));
	  exit (SYERR);
	}
	while (1) {
	  /* Parse line */
	  if (!new_line) {
	    rv = dlf_get_pair (in_file, par_name, par_str_val);
	  }
	  else {
	    new_line = 0;
	    rv = gp_rv;
	  }
	  if (rv <= 0) {
	    if (rv < 0) {
	      fprintf (stderr, "Bad log file format. Log file is %s\n", file_name);
	      exit (SYERR);
	    } else if (rv == 0) { /* End of file reached */
	      fclose (in_file);
	      fprintf (stdout, "%d messages have been transferred.\n", line_no - 1);
	      /* Connection to the server should be already closed */
	      exit(0);
	      /************************* PROGRAM END ******************************************/
	    }
	  } else if (rv == 1) { /* New line reached */
	      line_no++;  
	      new_line = 1;
	      /* Try to get new pair */
	      gp_rv =  dlf_get_pair (in_file, par_name, par_str_val);
	      if (gp_rv <= 0) last_message = 1; else last_message = 0;
	      rv = 0;
	      if (store_severity < 0 || log_message.severity == store_severity) {
		rv = dlf_send_log_message(&dst_host, &log_message, last_message);
	      }
	      dlf_delete_param_list (&log_message.param_list);
	      dlf_log_message_init(&log_message);
	      if (rv < 0) {
		fprintf (stderr, "Send message ERROR.\n");
		break;
	      }
	      else
		continue;
	  }
     
	  /* Search whether par_name is a keyword */
	  key = dlf_iskeyword(par_name);
	  rv = 0;
	  switch (key) {
	  case DLF_KEY_DATE:
	    /* Copy the date portion until '.' */
	    p = strchr (par_str_val, '.');
	    if (p == NULL) { /* We don't have microseconds */
	      n = strlen (par_str_val);
	      log_message.time_usec = 0;
	    }
	    else {
	      n = p - par_str_val;
	      if (dlf_isinteger(++p))
		log_message.time_usec = atoi (p);
	      else {
		rv = -1;
		break;
	      }
	    }
	    if (n != DLF_TIMESTRLEN) { /* Error in time value */
	      rv = -1;
	      serrno = EDLFLOGFORMAT;
	      break;
	    }
	    strncpy (log_message.time, par_str_val, n);
	    log_message.time[n] = '\0'; /* Terminate string */
	    break;
	  case DLF_KEY_HOST:
	    if (strlen(par_str_val) > CA_MAXHOSTNAMELEN) {
	      rv = -1;
	      serrno = EDLFLOGFORMAT;
	      break;
	    }
	    strcpy (log_message.hostname, par_str_val);
	    break;
	  case DLF_KEY_LVL:
	    for (i = 0; g_dlf_levels[i].lvl_id != 0 && strcmp(par_str_val, g_dlf_levels[i].lvl_name) != 0 ; i++);
	    log_message.severity = g_dlf_levels[i].lvl_id;
	    break;
	  case DLF_KEY_PRGN:
	    if (!dlf_isinteger( par_str_val )) {
	      rv = -1;
	      serrno = EDLFLOGFORMAT;
	      break;
	    }
	    log_message.facility_no = atoi(par_str_val);
	    break;
	  case DLF_KEY_PROG:
	    if (strlen(par_str_val) > DLF_MAXFACNAMELEN) {
	      rv = -1;
	      serrno = EDLFLOGFORMAT;
	      break;
	    }
	    break;
	  case DLF_KEY_PID:
	    if (!dlf_isinteger( par_str_val )) {
	      rv = -1;
	      serrno = EDLFLOGFORMAT;
	      break;
	    }
	    log_message.pid = atoi(par_str_val);
	    break;
	  case DLF_KEY_TID:
	    if (!dlf_isinteger( par_str_val )) {
	      rv = -1;
	      serrno = EDLFLOGFORMAT;
	      break;
	    }
	    log_message.cid = atoi(par_str_val);
	    break;
	  case DLF_KEY_MSGN:
	    if (!dlf_isinteger( par_str_val )) {
	      serrno = EDLFLOGFORMAT;
	      rv = -1;
	      break;
	    }
	    log_message.message_no = atoi(par_str_val);
	    break;
	  case DLF_KEY_MSG:
	    break;
	  case DLF_KEY_RQID:
	    rv = dlf_hex2uuid( par_str_val, log_message.request_id );
	    break;
	  case DLF_KEY_NSINV:
	    rv = dlf_hex2u64 ( par_str_val, &log_message.ns_invariant );
	    break;
	  case DLF_KEY_SRQID:
	    rv = dlf_hex2uuid( par_str_val, sub_request_id );
	    if (rv < 0)
	      break;
	    rv = dlf_add_subreq_id ( &log_message, sub_request_id );
	    break;
	  case DLF_KEY_TPVID:
	    rv = dlf_add_tpvid_parameter (&log_message, par_str_val);
	    break;
	  default:
	    /* Here we may get string, integer or floating point */
	    if (dlf_isinteger(par_str_val)) {
	      errno = 0; /* Clear error */
	      i64val = strtoll(par_str_val, &end_ptr, 10);
	      if (*end_ptr == '\0' && errno != ERANGE) {
		rv = dlf_add_int_parameter(&log_message, par_name, i64val);
	      }
	      else { /* Store it as a string */
		rv = dlf_add_str_parameter (&log_message, par_name, par_str_val);	      
	      }
	    }
	    else if (dlf_isfloat(par_str_val)) {
	      errno = 0; /* Clear error */
	      dval = strtod(par_str_val, &end_ptr);
	      if (*end_ptr == '\0' && errno != ERANGE) {
		rv = dlf_add_double_parameter(&log_message, par_name, dval);
	      }
	      else { /* Store it as a string */
		rv = dlf_add_str_parameter (&log_message, par_name, par_str_val);	      
	      }
	    }
	    else { /* It is a string */
	      rv = dlf_add_str_parameter (&log_message, par_name, par_str_val);	      
	    }
	    break;
	  }
	  if (rv < 0)
	    break;
	}
	fclose(in_file);
	fprintf (stderr, "Error occurred during processing of the file %s:-\n-Line number: %d  serror: %s\n",
		 file_name, line_no, sstrerror(serrno));
	exit (SYERR);
}

int DLL_DECL dlf_get_pair (in_file, par_name, par_str_val)
FILE *in_file;
char *par_name;
char *par_str_val;
{
  int c, d;
  char *p;
  int i;

  /* Skip spaces */
  while ((c = fgetc (in_file)) == ' ' || c == '\t');
  if ( c == '\n' ) return (1); /* End of line reached */
  if ( c == EOF ) return (0); /* End of file or read error */
  /* Get parameter name */
  ungetc (c, in_file);
  for (i = 0, p = par_name; (c = fgetc (in_file)) != '=' && c != '\n' && c != EOF && i < DLF_MAXPARNAMELEN; i++) p[i] = c;
  p[i] = '\0'; /* Terminate string */
  if (c != '=') return (-1); /* Parse error */
  /* Get parameter string */
  /* String may be delimited by \" or \' */
  /* Try to obtain delimiter (if any) */
  c = fgetc (in_file);
  if (c == EOF || c == '\n') return (-1); /* Parse error: unexpected end of line/file */
  if ( c == '"' || c == '\'') d = c; else {
    d = ' ';
    ungetc (c, in_file);
  }
  for (i = 0, p = par_str_val; (c = fgetc (in_file)) != d && c != '\n' && c != EOF && i < DLF_MAXSTRVALLEN; i++) p[i] = c;
  p[i] = '\0'; /* Terminate string */
  if (d == ' ') {
    ungetc (c, in_file);
  }
  return (2);
}

int DLL_DECL dlf_iskeyword(par_name)
const char *par_name;
{
  int i;
  int n;
  char *p;

  for (i = 0; keywords[i].key_id !=0; i++) {
    n = strlen(keywords[i].key_name) - 1; 
    if (keywords[i].key_name[n] == '*') {
      if (strncmp(keywords[i].key_name, par_name, n) == 0) {
	for (p = (char *)par_name + n; *p != '\0' && isdigit(*p); p++);
	if (p == par_name + strlen(par_name)) 
	  break;
	else
	  continue;
      }
    }
    else {
      if ( strcmp(keywords[i].key_name, par_name) == 0 ) break;
    }
  }
  return (keywords[i].key_id);
}

int DLL_DECL dlf_send_log_message(dst, log_message, last_message)
dlf_log_dst_t *dst;
dlf_log_message_t *log_message;
int last_message;
{
  char *buf;
  int buf_len;
  int rep_type;

  if (dlf_send_to_host(dst, log_message, 1, last_message) < 0)
    return (-1);
  /* Get reply */
  if (getrep (dst->descr, &buf, &buf_len, &rep_type) < 0) {
    return (-1);
  }
  if (!last_message) {
    if (rep_type != DLF_IRC) { /* Protocol error */
      /* Close connection */
      (void)netclose(dst->descr);
      return (-1);
    }
  }
  return (0);
}
